src/development/vortex/development/core/pipelines/training_pipeline.py

```diff
@@ -317,45 +322,37 @@ def run(self, save_model: bool = True) -> EasyDict:
                         if not metric_name in val_results:
                             val_res_key = ", ".join(list(val_results.keys()))
                             raise RuntimeError("'save_best_metric' value of ({}) is not found in validation "
                                 "result, choose either one of [{}]".format(metric_name, val_res_key))
-                        if best_metric[metric_name] < val_results[metric_name]:
-                            best_metric[metric_name] = val_results[metric_name]
+                        if self.best_metrics[metric_name] < val_results[metric_name]:
+                            self.best_metrics[metric_name] = val_results[metric_name]
                             model_fname = "{}-best-{}.pth".format(self.config.experiment_name, metric_name)
                             self._save_checkpoint(epoch, metrics=val_results, filename=model_fname)
 
                 print('epoch %s validation : %s' % (epoch, ', '.join(['{}:{:.4f}'.format(key, value) for key, value in val_results.items()])))
 
             ## save on best loss
-            if self.save_best_type and 'loss' in self.save_best_type and best_loss > loss.item() and save_model:
-                best_loss = loss.item()
-                metrics = None
-                if self.valid_for_validation:
-                    metrics = val_results
+            if self.save_best_type and 'loss' in self.save_best_type and \
+                    self.best_metrics['loss'] > loss.item() and save_model:
+                self.best_metrics['loss'] = loss.item()
                 model_fname = "{}-best-loss.pth".format(self.config.experiment_name)
-                self._save_checkpoint(epoch, metrics=metrics, filename=model_fname)
+                self._save_checkpoint(epoch, metrics=val_results, filename=model_fname)
 
             # Save on several epoch
             if self.save_epoch and ((epoch+1) % self.save_epoch == 0) and save_model:
-                metrics = None
-                if self.valid_for_validation:
-                    metrics = val_results
                 model_fname = "{}-epoch-{}.pth".format(self.config.experiment_name, epoch)
-                self._save_checkpoint(epoch, metrics=metrics, filename=model_fname)
-
-        # Save final weights on after all epochs finished
-        if save_model:
-            metrics = None
-            if self.valid_for_validation:
-                metrics = val_results
-            model_fname = "{}.pth".format(self.config.experiment_name)
-            model_path = self._save_checkpoint(last_epoch, metrics=metrics, filename=model_fname)
-            # Copy final weight from runs directory to experiment directory
-            shutil.copy(model_path, Path(self.experiment_directory))
+                self._save_checkpoint(epoch, metrics=val_results, filename=model_fname)
+
+            if save_model: ## save this epoch
+                model_fname = "{}.pth".format(self.config.experiment_name, epoch)
```

Comment:
I think better use more verbose name such as '{experiment-name}-last.pth rather than making it the same as final weight. And for better separability I prefer the old way to save final weight out of epoch iteration

Reply:
okay then.


> And for better separability I prefer the old way to save final weight out of epoch iteration  

there should be no difference in the final weight right? i think it is simpler this way

Reply:
oh you're right, avoid redundant weight

Reply:
so how is it then?

Reply:
yep keep it that way, but please update the naming for last epoch weight

Reply:
okay done with 254fd4a